Header updates

imjuangarcia opened the pull request:
## 📚 Context

This PR introduces a handful of design adjustments on the app header, that were requested internally by some folks (see Notion pages linked below).

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [ ] Feature
  - [ ] Refactoring
  - [x] Other, please describe: Design adjustments

## 🧠 Description of Changes

* Styled the top buttons (Share, Favorite, Menu, Sidebar) so that they have consistent paddings and font sizes throughout;
* Adds a different hover state on these buttons;
* The top header now has a semi-opaque background to improve legibility as you scroll;
* Now that the top header has this backdrop, we removed the background behind the running man animation, and aligned the `Running...` text to the baseline of the other buttons.

  - [ ] This is a breaking API change
  - [x] This is a visible (user-facing) change

**Revised:**

_Paddings and font sizes_
![Screen Shot 2022-03-11 at 08 45 00](https://user-images.githubusercontent.com/34423371/157860940-b729c85d-0971-4fe0-b653-694bb524808b.png)

_Hover states: Light mode_
![Screen Shot 2022-03-11 at 08 42 48](https://user-images.githubusercontent.com/34423371/157861321-1c1736d3-95e5-4ef6-9fe2-43b624b9f9b6.png)

_Hover states: Dark mode_
![Screen Shot 2022-03-11 at 08 43 17](https://user-images.githubusercontent.com/34423371/157861393-0f71ab84-b7f3-4f01-81ca-fd16355b64c1.png)

_Hover states: Custom theme (takes the primary color, and transparentizes it)_
![Screen Shot 2022-03-11 at 08 44 02](https://user-images.githubusercontent.com/34423371/157861501-58d438ab-ad90-410d-ade9-d4f7aec3ba90.png)

_Videos with the functionality for the sticky header_
* Dark mode: https://share.getcloudapp.com/Wnuy4WWX
* Light mode: https://share.getcloudapp.com/04uEenDy
* Custom theme: https://share.getcloudapp.com/9ZumrJpK

**Current:**

_Paddings and font sizes_
![Screen Shot 2022-03-11 at 08 44 49](https://user-images.githubusercontent.com/34423371/157860973-b08e0bce-ae70-43b3-85e4-a150d4dd46e8.png)

_Hover states: Light mode_
![Screen Shot 2022-03-11 at 08 42 27](https://user-images.githubusercontent.com/34423371/157861230-0f4d36c4-8cd9-4f23-85b4-0064f071726a.png)

_Hover states: Dark mode_
![Screen Shot 2022-03-11 at 08 43 04](https://user-images.githubusercontent.com/34423371/157861263-950b6aa0-1dbb-49ca-9059-e5107698123d.png)

_Previous header scrolling behavior_
![Screen Shot 2022-03-11 at 08 53 47](https://user-images.githubusercontent.com/34423371/157862228-ce0c5085-a497-47b2-8ff2-0cdfe381c65c.png)

## 🧪 Testing Done

- [x] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

* https://www.notion.so/streamlit/Header-buttons-should-be-same-size-fdb226ff5ff94758b614b2c8efd71246
* https://www.notion.so/streamlit/Header-button-hover-should-have-gray-background-ddc75c1481b64193a9dc0768374e4226

- **Issue**: Closes #XXXX

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

Commit:
Create new type for header buttons

Commit:
Functionality to change button styles based on selected theme

Commit:
Change style for hamburger menu based on selected theme

Commit:
Remove background color from status widget

Commit:
Remove console log statement

Commit:
Adjustments on statusWidget components

Commit:
Make top header translucent

Commit:
Style the sidebar buttons using the same style as the header ones

AnOctopus commented:
You will need to fix some linting issues (use `make jslint` to see/check them locally), and will probably need to update snapshots for e2e tests

Commit:
Fix JS linting errors

imjuangarcia commented:
@AnOctopus Thanks for letting me know! I fixed the linting errors on [f3fc963](https://github.com/streamlit/streamlit/pull/4496/commits/f3fc963147c73cdc417abfee1d8cb5624feb9eb6), but when trying to update the e2e test, I'm hung up on `Waiting for React app server to come online...`

I'm on an `amd64` architecture machine, could that be the issue? I'm getting the warning I'm attaching below on Docker desktop:

![Screen Shot 2022-03-11 at 15 32 39](https://user-images.githubusercontent.com/34423371/157930660-8b35c789-2b2f-4a5e-a5cb-dc09223e1aa5.png)

Thanks for the help!

vdonato commented:
> I'm on an `amd64` architecture machine, could that be the issue? I'm getting the warning I'm attaching below on Docker desktop

Hm, that's promising compared to where things were before (last time I checked, the snapshot testing image didn't even work on arm), but I guess it'll still be awhile before everything we use is ported to work on non-x86 architectures.

We haven't prioritized fixing the snapshot image because generally it's pretty rare that snapshots need to be updated, and when they are there are few enough of them that deleting existing ones and using CI to regenerate them has been a viable solution. It looks like there are too many snapshots to update for this PR, though, so I wouldn't recommend the CI method.

I still have an x86 machine with a dev env on it and can help update snapshots later today 🙂

imjuangarcia commented:
@vdonato That'd be awesome, thanks a ton! I tried a couple times after I replied on the PR, with different snapshots (`st_button`, `st_header`, `st_mainMenu`) but it failed on all of them. Appreciate your help!

vdonato commented on frontend/src/App.tsx:
```diff
@@ -1129,12 +1129,16 @@ export class App extends PureComponent<Props, State> {
                     stopScript={this.stopScript}
                     allowRunOnSave={allowRunOnSave}
+                    activeTheme={this.props.theme.activeTheme.name}
+                    baseTheme={this.props.theme.activeTheme.baseweb.name}
```
Hm, so I think we can do this more cleanly than plumbing these theme names around to a bunch of different places, which ends up getting quite verbose.

An alternative way of doing this that I think would be cleaner would be to:
1. define new `menuBtnHover` colors in `theme/lightTheme/themeColors.ts` and `theme/darkTheme/themeColors.ts` to be the colors that are currently defined in `getThemeStyle`
2. in `createTheme` (which lives in `theme/utils.ts`), overwrite `emotion.colors.menuBtnHover` to be a transparentized primary color if we're creating a custom theme
3. Use `theme.colors.menuBtnHover` directly in the new `StyledMenuBtn`'s hover state instead of the `getThemeStyle` function. This should let us do the same thing without plumbing around the new `activeTheme` and `baseTheme` props

Note that depending on how we handle some of the design decisions I asked about in another comment, we may not even have to do step 2.

imjuangarcia replied:
> Hm, so I think we can do this more cleanly than plumbing these theme names around to a bunch of different places, which ends up getting quite verbose.

Agreed! Removed on [a05426b](https://github.com/streamlit/streamlit/pull/4496/commits/a05426b54039cbe4b4dfc36973cdf0a6f3635551)

> Note that depending on how we handle some of the design decisions I asked about in another comment, we may not even have to do step 2.

I found a workaround for this, using the `transparentDarkenedBgMix60` computed property I mentioned below, and added it on [722a443](https://github.com/streamlit/streamlit/pull/4496/commits/722a443e7b282f1e066e7b20b5587473c9ebbb80)

tvst commented:
@imjuangarcia since Core is a bit weird with respect to its design system, can you make a wheel file and share with me and Jess before merging?

That's:
```
make frontend
make distribution
```

Then the wheel file will be a `.whl` file at `./lib/dist/`

vdonato commented on frontend/src/components/core/Header/styled-components.ts:
```diff
@@ -30,7 +30,12 @@ export const StyledHeader = styled.header<StyledHeaderProps>(
     right: theme.spacing.none,
-    height: 0,
+    height: "3.375rem",
+    background: `linear-gradient(0deg, ${transparentize(
+      theme.colors.bgColor,
+      0.5
+    )}, ${theme.colors.bgColor})`,
```
Hm, so the linear gradient here is introducing some tricky problems with our snapshot tests that we'll need to figure out how to resolve. I noticed a bunch of seemingly unrelated snapshots being updated when I was batch-updating them yesterday (in particular, most dataframe and table-related snapshots were failing). 

It turns out what's going on here is that the newly added gradient is being picked up and included in the snapshots that are being taken, which is causing false-positive snapshot failures. See the snapshot and video of the test run that I've attached below.

I think the fix for this will be to simply hide the header before we start taking snapshots of things so that the gradient doesn't end up appearing, but I'll have to think about this a bit to make sure that doing this doesn't have any unintended consequences (I can revisit helping out with this once my other comments are resolved).

Video of test run:
https://user-images.githubusercontent.com/3144420/158038786-9495d421-a630-43a8-91aa-b4d6aa196188.mp4

Diff showing expected snapshot and actual snapshot where you can see the gradient
![arrow-table-unstyled-dark diff](https://user-images.githubusercontent.com/3144420/158038870-d013b598-0342-4a9d-92c7-4060d37db693.png)

imjuangarcia replied:
Thank you for going the extra mile with this! I can't really speak on which would be the best solution, but I appreciate helping figure it out.

vdonato replied:
No problem! The snapshot situation has been far from ideal for awhile now, but we've been able to put off figuring out how to improve it given how rarely they get updated.

In case these design fixes across the core codebase are likely to be a much more common thing, let us know and we'll try to prioritize improving the process to be much more streamlined

vdonato commented on frontend/src/components/shared/Button/styled-components.ts:
```diff
@@ -71,6 +74,43 @@ function getSizeStyle(size: Size, theme: Theme): CSSObject {
   }
 }
 
+function getThemeStyle(
```
Note that I think that if we make the changes to simplify this change that I described in another PR, the need for this helper function will go away.

imjuangarcia replied:
That makes sense! I just asked the product folks to check if they are ok with the behavior you explained below. Even if they still prefer the transparentized background version for custom themes, I noticed we have a `transparentDarkenedBgMix60` variable we could use for that matter, so this fn wouldn't be necessary either way.

vdonato commented on frontend/src/components/shared/Button/styled-components.ts:
```diff
@@ -71,6 +74,43 @@ function getSizeStyle(size: Size, theme: Theme): CSSObject {
+          backgroundColor: theme.colors.gray90,
+        },
+      }
+    case "Custom Theme":
+      return {
+        "&:hover": {
+          backgroundColor: transparentize(theme.colors.primary, 0.5),
+        },
+      }
```
This is more of a design question than a question about the code itself, but do we want to switch to using a transparentized primary color for all custom themes? It would feel a bit weird to me if the color of the menu buttons changed after I made a custom theme that leaves most of the colors the same but only changes the `textColor` or `font` options.

Currently, the way that we pick some "auxiliary" colors (such as the colors in the `st.info`/`st.warning`/`st.error` components) for a custom theme is to pick the same colors that we use in light/dark mode based on the luminance of the background color (see [this comment](https://github.com/streamlit/streamlit/blob/02cbb2244d1617e8b417ab2a7bb817691b973efc/frontend/src/theme/utils.ts#L480) and the function it lives in for how this is done). I feel like doing something similar here may make sense.

imjuangarcia replied:
Thanks for explaining this! I appreciate it, as I'm totally new to the codebase. I brought this and the `transparentDarkenedBgMix60` options to a design meeting with the Product folks, and they preferred the latter, which I believe also addresses the point you mentioned here: 

> It would feel a bit weird to me if the color of the menu buttons changed after I made a custom theme that leaves most of the colors the same but only changes the textColor or font options.

As it's based on the `bgColor`, and not the `primary` color anymore. I added that on [722a443](https://github.com/streamlit/streamlit/pull/4496/commits/722a443e7b282f1e066e7b20b5587473c9ebbb80), but let me know if you have any additional input!

Commit:
Remove activeTheme and baseTheme props

Commit:
Remove unused props and use computed property for hover state

vdonato reviewed (commented):
🎉 This LGTM, but there's just a handful of things left to do before this is merge-able

1) I still need to sit down and figure out a way of making sure the gradient in the header doesn't interfere with snapshots. I don't have much time to work on this today, but I expect to be able to help out tomorrow
2) It looks like there are still a few places where `activeTheme` and `baseTheme` props are passed around, so those can be removed. Aside from this, though, I don't think there are any changes that need to be made to non-test code

Commit:
Remove remaining unused props

imjuangarcia commented:
>     1. I still need to sit down and figure out a way of making sure the gradient in the header doesn't interfere with snapshots. I don't have much time to work on this today, but I expect to be able to help out tomorrow

No rush! Thanks 🙏

>     2. It looks like there are still a few places where `activeTheme` and `baseTheme` props are passed around, so those can be removed. Aside from this, though, I don't think there are any changes that need to be made to non-test code

Sorry I missed those. They were all deleted on [c0c0d64](https://github.com/streamlit/streamlit/pull/4496/commits/c0c0d647b28814d2a94c0356cb14887277ce5945) now

tvst commented:
LGTM!

Just one nit: can you make the ":focus" state of all the header buttons use ":focus-visible" instead?

How: set `outline` to "none" on `:focus` state, and set the outline to what you actually want on `:focus-visible`, like this:
https://github.com/streamlit/streamlit/blob/d44b162909fb8adcae463172c78000029e5d2fef/frontend/src/components/shared/FullScreenWrapper/styled-components.ts#L61

Commit:
Fix on focus states for header buttons

imjuangarcia commented:
> Just one nit: can you make the ":focus" state of all the header buttons use ":focus-visible" instead?

Done on [10a6d58](https://github.com/streamlit/streamlit/pull/4496/commits/10a6d58adb4d6aa7923f045a9b158a72588cd3f2). I also removed the `&:not(:active)` rule, as it was taking precedence over the other pseudo-element ones.

vdonato reviewed (approved):
LGTM!

There's still a bit more to do (all on my end) before this is mergeable:
* merge https://github.com/streamlit/streamlit/pull/4515 once it's reviewed
* have the newly-added cypress command also hide the gradient/blur so that it can't interfere with snapshots
* Use the snapshot docker image to update all affected snapshots

I'll aim to wrap up the last required bits tomorrow -- sorry that getting this merged is taking a surprisingly long amount of time!

vdonato committed:
Merge branch 'develop' into header-updates

vdonato committed:
Remove some header styling in cy.prepForElementSnapshots

vdonato committed:
Update jest snapshots and fix some tests

Note this required some small name changes so that the new button type has a
name that matches what tests expect.

vdonato committed:
Call cy.prepForElementSnapshots in typography.spec.js

vdonato committed:
Fix last e2e tests and update snapshots

vdonato commented:
@imjuangarcia I think it'll be worth double-checking the changes that I made on your branch before clicking merge, but once you've done that we're good to go 🎉 

Here are the notable changes that I made:
* Updated all affected e2e snapshots (I just ran the full e2e test suite)
* Made some small tweaks to the e2e tests to get everything running. Most of these changes involved calling the new helper function I added in https://github.com/streamlit/streamlit/pull/4515
* Updated all `.tsx.snap` jest snapshots (note that some of these have gotten much more verbose since a package upgrade for reasons I don't fully understand yet, so I don't think it's worth reading any of these diffs)
* Changed a `styled-component` name from `StyledMenuBtn` to `StyledHeaderButton`, as well as the string name of the kind to `"header"`. This had to be done because our tests expect all of the `Styled*Button` components to have names following that pattern

Feel free to go ahead and click merge if everything looks good to you!

imjuangarcia commented:
@vdonato Thanks a ton for all your help with those! I can't really speak for the changes on the snapshot tests, but the visual changes look good on my end.

Seems like I'm not authorized to merge to the base branch, so feel free to do so yourself!

Thanks again 🙏

giswqs commented:
Is it possible to disable the top header semi-opaque background?

imjuangarcia commented:
Hi @giswqs! 

> Is it possible to disable the top header semi-opaque background?

Not at the moment, but we are changing its behavior to be solid instead of semi-opaque. I'll let you know when the PR with that change is merged!

giswqs commented:
@imjuangarcia Thanks. Some users dislike the semi-opaque header and asked me to disable it on my web app.

navigatio1 commented:
Agreed: The semi-opaque background for the top header makes scrolling legibility worse. It would be best to remove this or if there is a genuine need by some users, make it optional, not the default.
Many thanks for all the good work.

imjuangarcia commented:
Appreciate the thoughts, @navigatio1! We changed it to be a solid color on https://github.com/streamlit/streamlit/pull/4596, so it'll likely ship within the next release 🙂
